### ISSUE TITLE
Huge DO CASE statement causes an internal compiler error (insufficient stack)

### ISSUE THREAD

**Issue by cpyrgas** [X-Sharp/XSharpPublic#1214]
```
PROCEDURE DoTest()
DO CASE
CASE FALSE
CASE FALSE
CASE FALSE
// ... another 2000 times
END CASE

error XS9999: An internal compiler error has occurred: 'Insufficient stack to continue executing the program safely. This can happen from having too many functions on the call stack or function on the stack using too much stack space.',    at System.Runtime.CompilerServices.RuntimeHelpers.EnsureSufficientExecutionStack()
(...)
   at LanguageService.CodeAnalysis.XSharp.XSharpSyntaxWalker.Visit(SyntaxNode node)
   at LanguageService.CodeAnalysis.XSharp.XSharpSyntaxWalker.Visit(SyntaxNode node)
   at LanguageService.CodeAnalysis.XSharp.LocalBinderFactory.Visit(XSharpSyntaxNode syntax, Binder enclosing)
   at LanguageService.CodeAnalysis.XSharp.LocalBinderFactory.VisitPossibleEmbeddedStatement(StatementSyntax statement, Binder enclosing)
   at LanguageService.CodeAnalysis.XSharp.XSharpSyntaxWalker.Visit(SyntaxNode node)
   at LanguageService.CodeAnalysis.XSharp.XSharpSyntaxWalker.Visit(SyntaxNode node)
   at LanguageService.CodeAnalysis.XSharp.LocalBinderFactory.Visit(XSharpSyntaxNode syntax, Binder enclosing)
   at LanguageService.CodeAnalysis.XSharp.LocalBinderFactory.VisitPossibleEmbeddedStatement(StatementSyntax statement, Binder enclosing)
   at LanguageService.CodeAnalysis.XSharp.XSharpSyntaxWalker.Visit(SyntaxNode node)
   at LanguageService.CodeAnalysis.XSharp.XSharpSyntaxWalker.Visit(SyntaxNode node)
   at LanguageService.CodeAnalysis.XSharp.LocalBinderFactory.Visit(XSharpSyntaxNode syntax, Binder enclosing)
   at LanguageService.CodeAnalysis.XSharp.LocalBinderFactory.VisitPossibleEmbeddedStatement(StatementSyntax statement, Binder enclosing)
   at LanguageService.CodeAnalysis.XSharp.XSharpSyntaxWalker.Visit(SyntaxNode node)
   at LanguageService.CodeAnalysis.XSharp.XSharpSyntaxWalker.Visit(SyntaxNode node)
   at LanguageService.CodeAnalysis.XSharp.LocalBinderFactory.VisitBlock(BlockSyntax node)
   at LanguageService.CodeAnalysis.XSharp.XSharpSyntaxWalker.Visit(SyntaxNode node)
   at LanguageService.CodeAnalysis.XSharp.LocalBinderFactory.VisitMethodDeclaration(MethodDeclarationSyntax node)
   at LanguageService.CodeAnalysis.XSharp.XSharpSyntaxWalker.Visit(SyntaxNode node)
   at LanguageService.CodeAnalysis.XSharp.LocalBinderFactory.BuildMap(Symbol containingMemberOrLambda, SyntaxNode syntax, Binder enclosing, Action`2 binderUpdatedHandler)
   at LanguageService.CodeAnalysis.XSharp.ExecutableCodeBinder.ComputeBinderMap()
   at LanguageService.CodeAnalysis.XSharp.ExecutableCodeBinder.GetBinder(SyntaxNode node)
   at LanguageService.CodeAnalysis.XSharp.Binder.BindBlock(BlockSyntax node, DiagnosticBag diagnostics)
   at LanguageService.CodeAnalysis.XSharp.Binder.BindStatement(StatementSyntax node, DiagnosticBag diagnostics)
   at LanguageService.CodeAnalysis.XSharp.Binder.BindMethodBody(XSharpSyntaxNode declaration, BlockSyntax blockBody, ArrowExpressionClauseSyntax expressionBody, DiagnosticBag diagnostics)
   at LanguageService.CodeAnalysis.XSharp.Binder.BindMethodBody(XSharpSyntaxNode syntax, DiagnosticBag diagnostics)
   at LanguageService.CodeAnalysis.XSharp.MethodCompiler.BindMethodBody(MethodSymbol method, TypeCompilationState compilationState, DiagnosticBag diagnostics, VariableState nullableInitialState, ImportChain& importChain, Boolean& originalBodyNested, InitialState& forSemanticModel)
   at LanguageService.CodeAnalysis.XSharp.MethodCompiler.CompileMethod(MethodSymbol methodSymbol, Int32 methodOrdinal, ProcessedFieldInitializers& processedInitializers, SynthesizedSubmissionFields previousSubmissionFields, TypeCompilationState compilationState)

```

**Comment by RobertvanderHulst** [comment]
I will see if I can change the recursive loop to a for loop.
Otherwise, we will have to add a limit to the # of CASE statements in a DO CASE.

**Comment by cpyrgas** [comment]
I think a limit is fine...

**Comment by RobertvanderHulst** [comment]
Chris,
I can probably do this without limit by converting the code a bit differently like this (in C# )
```
DO CASE
CASE condition1
    statementlist1
CASE condition2 
    statementlist2
CASE condition3
    statementlist3
OTHERWISE
   statementlist4
ENDCASE
// rest of the code
```
this is now translated to a C# parse tree that looks like this:
```
if (condition1)
{
   statementlist1
}
else if (condition2)
{
   statementlist2
}
else if (condition3)
{
   statementlist3
}
else
{
   statementlist4
}
```
The compiler is iterating this and as you can see the else statements cause a lot of nesting.

I suggest changing the generated C# tree to

```
if (condition1)
{
   statementlist1
   goto compilergeneratedlabel;
}
if (condition2)
{
   statementlist2
   goto compilergeneratedlabel;
}
if (condition3)
{
   statementlist3
   goto compilergeneratedlabel;
}
else
{
   statementlist4
   goto compilergeneratedlabel;
}
compilergeneratedlabel:
// rest of the code

```
The goto statements in the last `if` can probably be removed, but for simplicity I will probably leave them. The backend will remove them anyway.

**Comment by cpyrgas** [comment]
Robert, if it does not cause any issue in the debugger active line jumping (I don't expect it does), then this sounds good, too!

**Comment by RobertvanderHulst** [comment]
Fixed in da811a6a65ab37d435e5b604c6eecbe91f32f719

**Comment by cpyrgas** [comment]
Confirmed fixed
